connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java

```diff
@@ -324,25 +324,31 @@ private PluginScanResult scanPluginPath(
         Collection<PluginDesc<T>> result = new ArrayList<>();
         for (Class<? extends T> plugin : plugins) {
             if (PluginUtils.isConcrete(plugin)) {
-                result.add(new PluginDesc<>(plugin, versionFor(plugin.newInstance()), loader));
+                result.add(new PluginDesc<>(plugin, versionFor(plugin, null), loader));
             } else {
                 log.debug("Skipping {} as it is not concrete implementation", plugin);
             }
         }
         return result;
     }
 
-    private <T> Collection<PluginDesc<T>> getServiceLoaderPluginDesc(Class<T> klass, ClassLoader loader) {
+    private <T> Collection<PluginDesc<T>> getServiceLoaderPluginDesc(Class<T> klass, ClassLoader loader)
+        throws InstantiationException, IllegalAccessException {
         ServiceLoader<T> serviceLoader = ServiceLoader.load(klass, loader);
         Collection<PluginDesc<T>> result = new ArrayList<>();
         for (T pluginImpl : serviceLoader) {
-            result.add(new PluginDesc<>((Class<? extends T>) pluginImpl.getClass(), versionFor(pluginImpl), loader));
+            result.add(new PluginDesc<>((Class<? extends T>) pluginImpl.getClass(), versionFor(pluginImpl.getClass(), pluginImpl), loader));
         }
         return result;
     }
 
-    private static <T>  String versionFor(T pluginImpl) {
-        return pluginImpl instanceof Versioned ? ((Versioned) pluginImpl).version() : "undefined";
+    private static <T> String versionFor(Class<? extends T> pluginKlass, T pluginImpl)
+        throws IllegalAccessException, InstantiationException {
+        if (pluginKlass.isAssignableFrom(Versioned.class)) {
+            return pluginImpl == null ? ((Versioned) pluginKlass.newInstance()).version() : ((Versioned) pluginImpl).version();
+        } else {
+            return "undefined";
+        }
```

Comment:
This logic is not exactly the most straightforward. What about something like this?
```
if (pluginKlass.isAssignableFrom(Versioned.class)) {
    Versioned versioned;
    if (pluginImpl != null) {
        versioned = (Versioned) pluginImpl;
    } else {
        versioned = (Versioned) pluginKlass.newInstance();
    }
    return versioned.version();
}
return "undefined";
```
or
```
if (pluginKlass.isAssignableFrom(Versioned.class)) {
    if (pluginImpl == null) {
        pluginImpl = pluginKlass.newInstance();
    }
    return ((Versioned) pluginImpl).version();
}
return "undefined";
```